### PR TITLE
Add social authentication options including - Apple, Microsoft, Facebook, and Google

### PR DESCRIPTION
**Pull request by eyatsenkoperpetio**
Add social authentication options including - Apple, Microsoft, Facebook, and Google

in config you need add:

SOCIAL_LOGINS:
ENABLED: true
APPLE_SIGNIN_ENABLED: true

MICROSOFT:
ENABLED: true
APP_ID: 'id'

GOOGLE:
ENABLED: true
GOOGLE_PLUS_KEY: 'id.apps.googleusercontent.com'
CLIENT_ID: 'id.apps.googleusercontent.com'

FACEBOOK:
ENABLED: true
FACEBOOK_APP_ID: 'id'
CLIENT_TOKEN: 'id'

### CODE REVIEW

**Comment by eyatsenkoperpetio** [comment]
Screenshot 2023-11-27 at 16 05 51
Screenshot 2023-11-27 at 16 05 41

**Comment by eyatsenkoperpetio** [comment]
in config you need add:

SOCIAL_LOGINS:
    ENABLED: true
    APPLE_SIGNIN_ENABLED: true

MICROSOFT:
  ENABLED: true
  APP_ID: 'id'

GOOGLE:
    ENABLED: true
    GOOGLE_PLUS_KEY: 'id.apps.googleusercontent.com'
    CLIENT_ID: 'id.apps.googleusercontent.com'

FACEBOOK:
    ENABLED: true
    FACEBOOK_APP_ID: 'id'
    CLIENT_TOKEN: 'id'

File: Core/Core/Network/AuthEndpoint.swift [hunk 1]
```diff
@@ -10,6 +10,8 @@ import Alamofire
 
 enum AuthEndpoint: EndPointType {
     case getAccessToken(username: String, password: String, clientId: String, tokenType: String)
+
```
Comment (mumer92) [inline Core/Core/Network/AuthEndpoint.swift:13]: please remove this extra line.

File: Core/Core/Network/AuthEndpoint.swift [hunk 2]
```diff
@@ -71,6 +77,14 @@ enum AuthEndpoint: EndPointType {
                 "asymmetric_jwt": true
             ]
             return .requestParameters(parameters: params, encoding: URLEncoding.httpBody)
+        case let .socialLogin(externalToken, _, clientId):
+            let params: [String: Encodable] = [
+                "client_id": clientId,
+                "token_type": "jwt",
+                "access_token": externalToken,
+                "asymmetric_jwt": true
```
Comment (mumer92) [inline Core/Core/Network/AuthEndpoint.swift:85]: now that we have a `token_type` coming from config, i think we should that value.

File: OpenEdX/AppDelegate.swift [hunk 3]
```diff
@@ -54,6 +61,31 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
...
+        if GIDSignIn.sharedInstance.handle(url) {
+            return true
+        }
+
+        if MSALPublicClientApplication.handleMSALResponse(
+            url,
+            sourceApplication: options[UIApplication.OpenURLOptionsKey.sourceApplication] as? String
+        ) {
+            return true
+        }
```
Comment (mumer92) [inline OpenEdX/AppDelegate.swift:84]: i think we should add check with the config, if a social auth is enabled, then handle the url, e.g.

```
if config.googleConfig.enabled {
        GIDSignIn.sharedInstance.handle(url)
}
```

**Review by mumer92 — CHANGES_REQUESTED** [review]
i think we dont need following, if any social provider is enabled, it should work.
```
SOCIAL_LOGINS:
ENABLED: true
```

File: Authorization/Authorization/Presentation/SocialSign/SocialSignViewModel.swift [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+import SwiftUI
+import Core
+import AuthenticationServices
+import FacebookLogin
+import GoogleSignIn
+import MSAL
+import Swinject
+
+enum Socials {
```
[the remark this answers is not in the document]
Comment (eyatsenkoperpetio) [reply Authorization/Authorization/Presentation/SocialSign/SocialSignViewModel.swift:16]: maybe SocialResult?

File: Authorization/Authorization/Presentation/Login/SignInViewModel.swift [hunk 5]
```diff
@@ -83,12 +91,92 @@ public class SignInViewModel: ObservableObject {
...
+    private func microsoftLogin(_ account: MSALAccount, _ token: String) {
+        socialLogin(
+            externalToken: token,
+            backend: "azuread-oauth2",
+            loginMethod: .microsoft
+        )
+    }
+
+    private func socialLogin(externalToken: String, backend: String, loginMethod: LoginMethod) {
+        Task { @MainActor in
```
[the remark this answers is not in the document]
Comment (eyatsenkoperpetio) [reply Authorization/Authorization/Presentation/Login/SignInViewModel.swift:151]: Screenshot 2023-11-28 at 11 01 11

File: Authorization/Authorization/Presentation/Registration/SignUpViewModel.swift [hunk 6]
```diff
@@ -108,7 +127,74 @@ public class SignUpViewModel: ObservableObject {
...
+
+    @MainActor
+    private func social(result: Socials) {
+        switch result {
+        case .apple(let credential):
+            appleLogin(credential)
+        case .facebook(let loginManagerLoginResult):
+            facebookLogin(loginManagerLoginResult)
+        case .google(let gIDSignInResult):
+            googleLogin(gIDSignInResult)
```
[the remark this answers is not in the document]
Comment (eyatsenkoperpetio) [reply Authorization/Authorization/Presentation/Registration/SignUpViewModel.swift:147]: Screenshot 2023-11-28 at 11 04 17

[the remark this answers is not in the document]
Comment (eyatsenkoperpetio) [reply Authorization/Authorization/Presentation/Registration/SignUpViewModel.swift:147]: because switch condition is result, and I show type of result

File: Authorization/Authorization/Presentation/Login/SignInViewModel.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (eyatsenkoperpetio) [reply Authorization/Authorization/Presentation/Login/SignInViewModel.swift:151]: if I add to method I need add to all places when I use this method, if inside Task only one place

**Comment by eyatsenkoperpetio** [comment]
Screenshot 2023-11-28 at 12 24 18

added SocialSignView_Previews

File: Authorization/Authorization/Presentation/Login/SignInViewModel.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (mumer92) [reply Authorization/Authorization/Presentation/Login/SignInViewModel.swift:151]: i think we should explicitly mark functions as `@MainActor` so we have clear idea that these functions are sync in nature
e.g.

```
@MainActor
    func sign(with result: Result<SocialResult, Error>) {
        result.success(social)
        result.failure { error in
            errorMessage = error.localizedDescription
        }
    }
    
    @MainActor
    private func social(result: SocialResult) {
        switch result {
        case .apple(let appleCredentials):
            appleLogin(appleCredentials, backend: result.backend)
        case .facebook:
            facebookLogin(backend: result.backend)
        case .google(let gIDSignInResult):
            googleLogin(gIDSignInResult, backend: result.backend)
        case .microsoft(_, let token):
            microsoftLogin(token, backend: result.backend)
        }
    }
    
    @MainActor
    private func appleLogin(_ credentials: AppleCredentials, backend: String) {
        socialLogin(
            externalToken: credentials.token,
            backend: backend,
            loginMethod: .apple
        )
    }

    @MainActor
    private func facebookLogin(backend: String) {
        guard let currentAccessToken = AccessToken.current?.tokenString else {
            return
        }
        socialLogin(
            externalToken: currentAccessToken,
            backend: backend,
            loginMethod: .facebook
        )
    }
    
    @MainActor
    private func googleLogin(_ result: GIDSignInResult, backend: String) {
        socialLogin(
            externalToken: result.user.accessToken.tokenString,
            backend: backend,
            loginMethod: .google
        )
    }
    
    @MainActor
    private func microsoftLogin(_ token: String, backend: String) {
        socialLogin(
            externalToken: token,
            backend: backend,
            loginMethod: .microsoft
        )
    }

    @MainActor
    private func socialLogin(externalToken: String, backend: String, loginMethod: LoginMethod) {
        Task {
            isShowProgress = true
            do {
                let user = try await interactor.login(externalToken: externalToken, backend: backend)
                analytics.setUserID("\(user.id)")
                analytics.userLogin(method: loginMethod)
                router.showMainOrWhatsNewScreen()
            } catch let error {
                isShowProgress = false
                if let validationError = error.validationError,
                   let value = validationError.data?["error_description"] as? String {
                    errorMessage = value
                } else if case APIError.invalidGrant = error {
                    errorMessage = CoreLocalization.Error.invalidCredentials
                } else if error.isInternetError {
                    errorMessage = CoreLocalization.Error.slowOrNoInternetConnection
                } else {
                    errorMessage = CoreLocalization.Error.unknownError
                }
            }
        }
    }

```

**Comment by mumer92** [comment]
@eyatsenkoperpetio i see in config we have `Apple` under `SOCIAL_LOGINS`
like this
```
SOCIAL_LOGINS:
            ENABLED: true
            APPLE_SIGNIN_ENABLED: true
```

how about remove `SOCIAL_LOGINS` and make separate for `Apple`?
in this way we dont have to relay on `SOCIAL_LOGINS` dictionary. 

Also can you update `ConfigTests` with the new dictionary and add test cases there?

File: Authorization/Authorization/Presentation/SocialSign/SocialSignViewModel.swift [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+}
+
+final public class SocialSignViewModel: ObservableObject {
+
+    // MARK: - Properties -
+
+    private var onSigned: ((Result<SocialResult, Error>) -> Void)
+
+    init(
+        config: ConfigProtocol = Container.shared.resolve(ConfigProtocol.self) ?? ConfigMock(),
```
Comment (mumer92) [inline Authorization/Authorization/Presentation/SocialSign/SocialSignViewModel.swift:43]: i think we should not provide default value from here in constructor, we should explicitly provide the config from outside.

File: Authorization/Authorization/Presentation/SocialSign/SocialSignViewModel.swift [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+        onSigned: @escaping (Result<SocialResult, Error>) -> Void
+    ) {
+        self.config = config
+        self.onSigned = onSigned
+    }
+
+    let  config: ConfigProtocol
```
Comment (mumer92) [inline Authorization/Authorization/Presentation/SocialSign/SocialSignViewModel.swift:50]:
```suggestion
    let config: ConfigProtocol
```

File: Authorization/Authorization/Presentation/SocialSign/SocialSignViewModel.swift [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+    private let appleSingInProvider: AppleSingInProvider  = .init()
+    private let googleSingInProvider: GoogleSingInProvider = .init()
+    private let facebookSingInProvider: FacebookSingInProvider = .init()
+    private let microsoftSingInProvider: MicrosoftSingInProvider = .init()
+
+    private var topViewController: UIViewController? {
+        UIApplication.topViewController()
+    }
+
+    // MARK: - Public Intens -
```
Comment (mumer92) [inline Authorization/Authorization/Presentation/SocialSign/SocialSignViewModel.swift:61]: can you change the `-` from the comments throughout? e.g. `// MARK: - Public Intens -` -> `// MARK: - Public Intens`

**Review by mumer92 — CHANGES_REQUESTED** [review]

File: Authorization/Authorization/Presentation/SocialSign/SocialSignViewModel.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (mumer92) [reply Authorization/Authorization/Presentation/SocialSign/SocialSignViewModel.swift]: `SocialAuthResult` would seem more appropriate, what do you think?

[the remark this answers is not in the document]
Comment (mumer92) [reply Authorization/Authorization/Presentation/SocialSign/SocialSignViewModel.swift]: also how about moving this enum out to a separate file?

File: Authorization/Authorization/Presentation/SocialSign/SocialSignViewModel.swift (hunk 9 above)
Comment (eyatsenkoperpetio) [reply Authorization/Authorization/Presentation/SocialSign/SocialSignViewModel.swift:61]: why?

**Review by mumer92 — APPROVED** [review]

**Comment by mumer92** [comment]
@saeedbashir please take a look into this.

**Review by saeedbashir — CHANGES_REQUESTED** [review]
@eyatsenkoperpetio Here is some initial feedback on testing the app. I'll review the code once the feature flow will be as per the intended behavior.
- Please add a proper PR description for clarity on what this PR is doing.
- How about the social auth view will appear on successful loading of registration form?
- In case of registration form fetching failure, there isn't any identification that the form is incomplete. There should be a full-screen error or maybe an error and a reload button to again fetch the registration form. (You may create a follow-up task for this.)
- It is required from `Apple` that if the app is providing any third-party social auth, it should also provide `Apple Signin`, so if any other social auth provider is also enabled. If any other third-party social auth is enabled, Apple Signin should also be enabled, regardless of the value of the APPLE_SIGNIN config.
- On canceling the social auth flow on the confirmation alert, the app is showing an error in the case of Google and Microsoft but not in the case of Facebook, and it would be nice if we could use the same error for all providers (in case of cancellation).
- Apple Signin throughing an error on click for me. What'm I missing here?
- The size of `Sign In` button is not aligned with the other fields in the iPad, as you are working here, could you please change its size and match it with other fields and social auth buttons? Same goes with the `Create account` button on register screen.
- In case, user is not registered with social auth and tried to sign in, on successful authentication, the app is showing a generic message, `something went wrong`.  I guess we should show a proper message and let the user know about the next steps.
- On registration page, after successful authentication with social auth, you might need to hit the `user_api/v1/account/registration/` again to get the auto-created name and username and hide the password field because a password isn't required in case of social auth registration (You may see  flow on the edX app for reference).

**Comment by eyatsenkoperpetio** [comment]
@saeedbashir  Apple Signin throughing an error on click for me. What'm I missing here? 

Screenshot 2023-11-30 at 15 12 14

**Comment by eyatsenkoperpetio** [comment]
It seems Apple Sign In capability entitlement is missing and thats why Apple Sign in isn't working

added but,  OpenEdx will not use social auth

**Comment by miankhalid** [comment]
@eyatsenkoperpetio I've discussed the matter with @saeedbashir and seems like the issue was with the implementation and is now fixed in your recent commits.

File: Authorization/Authorization/Presentation/Registration/SignUpViewModel.swift [hunk 10]
```diff
@@ -108,7 +123,174 @@ public class SignUpViewModel: ObservableObject {
...
+            facebookLogin(backend: result.backend, account: account, loginMethod: .facebook)
+        case .google(let gIDSignInResult):
+            googleLogin(gIDSignInResult, backend: result.backend, loginMethod: .google)
+        case .microsoft(let account, let token):
+            microsoftLogin(token, backend: result.backend, account: account, loginMethod: .microsoft)
+        }
+    }
+
+    @MainActor
+    private func appleLogin(
```
Comment (saeedbashir) [inline Authorization/Authorization/Presentation/Registration/SignUpViewModel.swift:167]: Isn't it a bit odd to use `Login` in the register?

File: Authorization/Authorization/Presentation/SocialSign/SocialSignView.swift [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+
+    init(
+        signType: SignType = .signIn,
+        viewModel: SocialSignViewModel
+    ) {
+        self._viewModel = .init(wrappedValue: viewModel)
+        self.signType = signType
+    }
+
+    enum SignType {
```
Comment (saeedbashir) [inline Authorization/Authorization/Presentation/SocialSign/SocialSignView.swift:25]: How about renaming it AuthType or SocialAuthType?

File: Authorization/Authorization/Presentation/SocialSign/SocialSignView.swift [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+        case signIn
+        case register
+    }
+    var signType: SignType = .signIn
+
+    private var title: String {
+        switch signType {
+        case .signIn:
+            return AuthLocalization.signInWith
```
Comment (saeedbashir) [inline Authorization/Authorization/Presentation/SocialSign/SocialSignView.swift:34]: In the latest xcode, the return is redundant, and it's also being used at multiple places in the codebase. So you can skip the return here.

File: Authorization/Authorization/Presentation/SocialSign/SocialSignViewModel.swift [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+        completion: @escaping (Result<SocialResult, Error>) -> Void
+    ) {
+        self.config = config
+        self.completion = completion
+    }
+
+    private let appleSingInProvider: AppleSingInProvider  = .init()
+    private let googleSingInProvider: GoogleSingInProvider = .init()
+    private let facebookSingInProvider: FacebookSingInProvider = .init()
+    private let microsoftSingInProvider: MicrosoftSingInProvider = .init()
```
Comment (saeedbashir) [inline Authorization/Authorization/Presentation/SocialSign/SocialSignViewModel.swift:54]: How about declaring them as `lazy` because it's not sure all of them are used together all the time?

File: Authorization/Authorization/Presentation/SocialSign/SocialSignViewModel.swift [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+
+    var isMicrosoftEnabled: Bool {
+        config.microsoft.enabled
+    }
+
+    var isAppleSignInEnabled: Bool {
+        if isFaceboolEnabled ||
+            isGoogleEnabled ||
+            isMicrosoftEnabled {
+            return true
```
Comment (saeedbashir) [inline Authorization/Authorization/Presentation/SocialSign/SocialSignViewModel.swift:78]: Can you please add a comment explaining why the `isAppleSignInEnabled` is true in case any other social auth provider is enabled?

File: Authorization/Authorization/Presentation/SocialSign/SocialSignViewModel.swift [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+
+    var isGoogleEnabled: Bool {
+        config.google.enabled
+    }
+
+    var isMicrosoftEnabled: Bool {
+        config.microsoft.enabled
+    }
+
+    var isAppleSignInEnabled: Bool {
```
Comment (saeedbashir) [inline Authorization/Authorization/Presentation/SocialSign/SocialSignViewModel.swift:74]: Although it's fine, but how about omitting `is` from names, making it like facebookEnabled? it will make it more simple/

File: Authorization/Authorization/en.lproj/Localizable.strings [hunk 16]
```diff
@@ -16,22 +16,35 @@
...
+"ERROR.AUTH_PROVIDER" = "This %@ account is not linked with any %@ account. Please register.";
 "ERROR.INVALID_EMAIL_ADDRESS_OR_USERNAME" = "Invalid email or username";
-"ERROR.INVALID_PASSWORD_LENGTH" = "Invalid password length";
 
 "SIGN_UP.TITLE" = "Sign up";
 "SIGN_UP.SUBTITLE" = "Create new account.";
 "SIGN_UP.CREATE_ACCOUNT_BTN" = "Create account";
 "SIGN_UP.HIDE_FIELDS" = "Hide optional Fields";
 "SIGN_UP.SHOW_FIELDS" = "Show optional Fields";
+"SIGN_UP.SUCCESS_SIGNEDIN_LABEL" = "You've successfully signed in.";
```
Comment (saeedbashir) [inline Authorization/Authorization/en.lproj/Localizable.strings:28]: how about renaming it to `SIGN_UP.SUCCESS_SIGNIN_LABEL` and the same will go for sublabel.

File: Authorization/Authorization/en.lproj/Localizable.strings [hunk 17]
```diff
@@ -16,22 +16,35 @@
...
+"GOOGLE" = "Google";
+"FACEBOOK" = "Facebook";
+"MICROSOFT" = "Microsoft";
+"OR" = "Or";
+
 "STARTUP.INFO_MESSAGE" = "Courses and programs from the world's best universities in your pocket.";
 "STARTUP.SEARCH_TITLE" = "What do you want to learn?";
 "STARTUP.SEARCH_PLACEHOLDER" = "Search our 3000+ courses";
 "STARTUP.EXPLORE_ALL_COURSES" = "Explore all courses";
+
```
Comment (saeedbashir) [inline Authorization/Authorization/en.lproj/Localizable.strings:50]: Extra new line.

File: Core/Core/AvoidingHelpers/Error/CustomError.swift [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+//  Core
+//
+//  Created by Eugene Yatsenko on 10.10.2023.
+//
+
+import Foundation
+
+public enum CustomError: Error {
+    case error(text: String)
+    case socialSignCanceled
```
Comment (saeedbashir) [inline Core/Core/AvoidingHelpers/Error/CustomError.swift:12]: How about renaming it to `socialAuthCanceled`?

File: Core/Core/AvoidingHelpers/Providers/AppleSingInProvider.swift [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+//
+
+import Foundation
+import AuthenticationServices
+import Swinject
+
+public struct AppleCredentials: Codable {
+    public var name: String
+    public var email: String
+    public var birthYear: String? = nil
```
Comment (saeedbashir) [inline Core/Core/AvoidingHelpers/Providers/AppleSingInProvider.swift:15]: Redundant Optional Initialization Violation warning.

Comment (saeedbashir) [inline Core/Core/AvoidingHelpers/Providers/AppleSingInProvider.swift:15]: I guess we are not using birth year in the registration process so you can omit it.

File: Core/Core/Configuration/Config/Config.swift [hunk 20]
```diff
@@ -13,8 +13,14 @@ public protocol ConfigProtocol {
     var tokenType: TokenType { get }
     var feedbackEmail: String { get }
     var appStoreLink: String { get }
+    var platformName: String { get }
+    var socialLoginEnabled: Bool { get }
```
Comment (saeedbashir) [inline Core/Core/Configuration/Config/Config.swift:17]: Can you please move this into FeaturesConfig?

File: Core/Core/Configuration/Config/FacebookConfig.swift [hunk 21]
```diff
@@ -0,0 +1,38 @@
...
+    case appID = "FACEBOOK_APP_ID"
+    case clientToken = "CLIENT_TOKEN"
+}
+
+public final class FacebookConfig: NSObject {
+    public var enabled: Bool = false
+    public var appID: String?
+    public var clientToken: String?
+
+    public var requiredKeysAvailable: Bool {
```
Comment (saeedbashir) [inline Core/Core/Configuration/Config/FacebookConfig.swift:21]: It can be private

File: Core/Core/Configuration/Config/FirebaseConfig.swift [hunk 22]
```diff
@@ -97,9 +97,9 @@ public class FirebaseConfig: NSObject {
     }
 }
 
-private let key = "FIREBASE"
+private let firebaseKey = "FIREBASE"
 extension Config {
     public var firebase: FirebaseConfig {
-        return FirebaseConfig(dictionary: self[key] as? [String: AnyObject] ?? [:])
+        FirebaseConfig(dictionary: self[firebaseKey] as? [String: AnyObject] ?? [:])
```
Comment (saeedbashir) [inline Core/Core/Configuration/Config/FirebaseConfig.swift:103]: Is this change required because it doesn't seem relevant?

File: Core/Core/Configuration/Config/GoogleConfig.swift [hunk 23]
```diff
@@ -0,0 +1,37 @@
...
+    case googlePlusKey = "GOOGLE_PLUS_KEY"
+    case clientID = "CLIENT_ID"
+}
+
+public final class GoogleConfig: NSObject {
+    public var enabled: Bool = false
+    public var googlePlusKey: String?
+    public var clientID: String?
+
+    public var requiredKeysAvailable: Bool {
```
Comment (saeedbashir) [inline Core/Core/Configuration/Config/GoogleConfig.swift:21]: It can be private

File: Core/Core/Configuration/Config/MicrosoftConfig.swift [hunk 24]
```diff
@@ -0,0 +1,35 @@
...
+private enum MicrosoftKeys: String {
+    case enabled = "ENABLED"
+    case appID = "APP_ID"
+}
+
+public final class MicrosoftConfig: NSObject {
+    public var enabled: Bool = false
+    public var appID: String?
+
+    public var requiredKeysAvailable: Bool {
```
Comment (saeedbashir) [inline Core/Core/Configuration/Config/MicrosoftConfig.swift:19]: It can be private

File: Core/Core/Extensions/CGColorExtension.swift [hunk 25]
```diff
@@ -31,3 +31,9 @@ public extension Color {
         return UIColor(self)
     }
 }
+
+public extension UIColor {
+    var sui: Color {
```
Comment (saeedbashir) [inline Core/Core/Extensions/CGColorExtension.swift:36]: Is it for SwiftUIColor? if so how about giving a proper/descriptive name?

File: Core/Core/Network/AuthEndpoint.swift [hunk 26]
```diff
@@ -21,6 +22,8 @@ enum AuthEndpoint: EndPointType {
         switch self {
         case .getAccessToken:
             return "/oauth2/access_token"
+        case let .socialLogin(_, backend, _, _):
```
Comment (saeedbashir) [inline Core/Core/Network/AuthEndpoint.swift:25]: I'm not sure but can be given the name `exchangeToken`, like the name `getAccessToken` given to login/signin.

File: Core/Core/Network/AuthEndpoint.swift [hunk 27]
```diff
@@ -40,6 +43,8 @@ enum AuthEndpoint: EndPointType {
         switch self {
         case .getAccessToken:
             return .post
+        case .socialLogin:
+            return .post
```
Comment (saeedbashir) [inline Core/Core/Network/AuthEndpoint.swift:47]: How about using an OR case condition with `getAccessToken`?

File: Core/Core/View/Base/LabelButton.swift [hunk 28]
```diff
@@ -0,0 +1,74 @@
+//
+//  LabelButton.swift
+//  Core
+//
+//  Created by Eugene Yatsenko on 10.10.2023.
+//
+
+import SwiftUI
+
+public struct LabelButton: View {
```
Comment (saeedbashir) [inline Core/Core/View/Base/LabelButton.swift:10]: Can you rename it to `SocailAuthButton` because we can't say we can generalize this button style for other parts of the app?

File: Core/CoreTests/Configuration/ConfigTests.swift [hunk 29]
```diff
@@ -32,7 +32,24 @@ class ConfigTests: XCTestCase {
...
+            "FACEBOOK_APP_ID": "facebookAppId",
+            "CLIENT_TOKEN": "client_token"
+        ],
+        "MICROSOFT": [
+            "ENABLED": true,
+            "APP_ID": "appId"
+        ],
+        "APPLE_SIGNIN": [
+            "ENABLED": true
+        ]
```
Comment (saeedbashir) [inline Core/CoreTests/Configuration/ConfigTests.swift:52]: Can you please add tests for these values as well?

File: OpenEdX/AppDelegate.swift [hunk 30]
```diff
@@ -31,15 +34,23 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
...
     ) -> Bool {
-        
         initDI()
         
-        if let config = Container.shared.resolve(ConfigProtocol.self),
-           let configuration = config.firebase.firebaseOptions {
-            FirebaseApp.configure(options: configuration)
-            Crashlytics.crashlytics().setCrashlyticsCollectionEnabled(true)
+        if let config = Container.shared.resolve(ConfigProtocol.self) {
+
```
Comment (saeedbashir) [inline OpenEdX/AppDelegate.swift:40]: extra new line.

File: Core/Core/AvoidingHelpers/Providers/AppleSingInProvider.swift [hunk 31]
```diff
@@ -0,0 +1,101 @@
...
+            return
+        }
+
+        var storage = Container.shared.resolve(CoreStorage.self)
+
+        let givenName = credentials.fullName?.givenName ?? storage?.appleSignGivenName ?? ""
+        let familyName = credentials.fullName?.familyName ?? storage?.appleSignFamilyName ?? ""
+
+        let email = credentials.email ?? storage?.appleSignEmail ?? ""
+        var name: String =  "\(givenName) \(familyName)"
```
Comment (saeedbashir) [inline Core/Core/AvoidingHelpers/Providers/AppleSingInProvider.swift:52]: Can we just save the formatted name instead of saving separately saving family and given name?

File: Core/Core/AvoidingHelpers/Providers/AppleSingInProvider.swift [hunk 32]
```diff
@@ -0,0 +1,101 @@
...
+            storage?.appleSignFamilyName = familyName
+        }
+
+        if storage?.appleSignGivenName == nil, !givenName.isEmpty {
+            storage?.appleSignGivenName = givenName
+        }
+
+        if storage?.appleSignEmail == nil {
+            storage?.appleSignEmail = email
+        }
```
Comment (saeedbashir) [inline Core/Core/AvoidingHelpers/Providers/AppleSingInProvider.swift:64]: I guess Apple returns email everytime, so there is no need to save the email.

**Review by saeedbashir — CHANGES_REQUESTED** [review]

**Comment by saeedbashir** [comment]
@eyatsenkoperpetio Tests are also failing. Please take a look into it as well.

File: Authorization/Authorization/Presentation/SocialSign/SocialSignViewModel.swift (hunk 13 above)
Comment (volodymyr-chekyrta) [inline Authorization/Authorization/Presentation/SocialSign/SocialSignViewModel.swift:54]: We should not initialize dependencies manually, this is a violation of the SOLID principles.
We have to provide dependencies to the initializer block via DI.

You probably won't be able to write tests for this class if you create dependencies manually.

File: Authorization/Authorization/Presentation/SocialSign/SocialSignViewModel.swift [hunk 33]
```diff
@@ -0,0 +1,150 @@
...
+        return config.appleSignIn.enable
+    }
+
+    // MARK: - Public Intens
+
+    func signInWithApple() {
+        appleSingInProvider.request { [weak self] result in
+            guard let self = self else {
+                return
+            }
```
Comment (volodymyr-chekyrta) [inline Authorization/Authorization/Presentation/SocialSign/SocialSignViewModel.swift:89]: From Swift 5.7 `guard let self else { return }` is enough.

File: Core/Core/AvoidingHelpers/Providers/AppleSingInProvider.swift [hunk 34]
```diff
@@ -0,0 +1,101 @@
...
+    public func authorizationController(
+        controller: ASAuthorizationController,
+        didCompleteWithAuthorization authorization: ASAuthorization
+    ) {
+        guard let credentials = authorization.credential as? ASAuthorizationAppleIDCredential else {
+            completion?(.failure(CustomError.unknownError))
+            return
+        }
+
+        var storage = Container.shared.resolve(CoreStorage.self)
```
Comment (volodymyr-chekyrta) [inline Core/Core/AvoidingHelpers/Providers/AppleSingInProvider.swift:46]: It should be injected into the initializer.
Entity shouldn't manage dependency and call DI itself.

File: Core/Core/AvoidingHelpers/Error/CustomError.swift [hunk 35]
```diff
@@ -0,0 +1,27 @@
+//
+//  CustomError.swift
+//  Core
+//
+//  Created by Eugene Yatsenko on 10.10.2023.
+//
+
+import Foundation
+
+public enum CustomError: Error {
```
Comment (volodymyr-chekyrta) [inline Core/Core/AvoidingHelpers/Error/CustomError.swift:10]: Should be a more meaningful name for better understanding the area of this error.
Mb AuthError if it's' connected to the Auth logic.

File: Authorization/Authorization/Presentation/SocialSign/SocialSignViewModel.swift (hunk 13 above)
Comment (volodymyr-chekyrta) [reply Authorization/Authorization/Presentation/SocialSign/SocialSignViewModel.swift:54]: After discussion decided to leave it as is for simplicity

File: Authorization/Authorization/en.lproj/Localizable.strings [hunk 36]
```diff
@@ -16,21 +16,34 @@
...
 
 "FORGOT.TITLE"= "Forgot password";
 "FORGOT.DESCRIPTION" = "Please enter your log-in or recovery email address below and we will send you an email with instructions.";
 "FORGOT.REQUEST" = "Reset password";
 "FORGOT.CHECK_TITLE" = "Check your email";
 "FORGOT.CHECK_Description" = "We have sent a password recover instructions to your email ";
 
+
+"SIGN_IN_WITH" = "Sign in with";
+"SIGN_IN_REGISTER" = "Register with";
```
Comment (saeedbashir) [inline Authorization/Authorization/en.lproj/Localizable.strings:40]: How about renaming it to REGISTER_WITH?

File: Authorization/Authorization/en.lproj/Localizable.strings [hunk 37]
```diff
@@ -16,21 +16,34 @@
 "SIGN_IN.LOG_IN_BTN" = "Sign in";
 
 "ERROR.INVALID_EMAIL_ADDRESS" = "Invalid email address";
+"ERROR.INVALID_PASSWORD_LENGHT" = "Invalid password lenght";
+"ERROR.AUTH_PROVIDER" = "This %@ account is not linked with any %@ account. Please register.";
```
Comment (saeedbashir) [inline Authorization/Authorization/en.lproj/Localizable.strings:20]: How about renaming it to `ERROR.ACCOUNT_NOT_LINKED` OR `ERROR.ACCOUNT_NOT_REGISTERED`?

File: Authorization/Authorization/en.lproj/Localizable.strings [hunk 38]
```diff
@@ -16,21 +16,34 @@
...
 "ERROR.INVALID_EMAIL_ADDRESS_OR_USERNAME" = "Invalid email or username";
-"ERROR.INVALID_PASSWORD_LENGTH" = "Invalid password length";
+"ERROR.ACCOUNT_DISABLED" = "Your account is disabled. Please contact customer support for assistance.";
```
Comment (saeedbashir) [inline Authorization/Authorization/en.lproj/Localizable.strings:22]: Although it's fine but `ERROR.DISABLED_ACCOUNT` sounds good.

File: Core/Core/Network/AuthEndpoint.swift [hunk 39]
```diff
@@ -10,6 +10,7 @@ import Alamofire
 
 enum AuthEndpoint: EndPointType {
     case getAccessToken(username: String, password: String, clientId: String, tokenType: String)
+    case getEchangeAccessToken(externalToken: String, backend: String, clientId: String, tokenType: String)
```
Comment (saeedbashir) [inline Core/Core/Network/AuthEndpoint.swift:13]: How about renaming it to `exchangeAccessToken`?

File: Core/Core/Providers/SocialAuth/AppleAuthProvider.swift [hunk 40]
```diff
@@ -0,0 +1,103 @@
...
+            name = pncf.string(from: fullName)
+        }
+
+        let email = credentials.email ?? storage?.appleSignEmail ?? ""
+
+        if !name.isEmpty {
+            storage?.appleSignFullName = name
+        }
+
+        if storage?.appleSignEmail == nil {
```
Comment (saeedbashir) [inline Core/Core/Providers/SocialAuth/AppleAuthProvider.swift:64]: The value of `storage?.appleSignEmail` is an empty string so the email is not being saved into storage for later ussage.

File: Core/Core/Providers/SocialAuth/FacebookAuthProvider.swift [hunk 41]
```diff
@@ -0,0 +1,63 @@
+//
+//  FacebookSingInProvider.swift
```
Comment (saeedbashir) [inline Core/Core/Providers/SocialAuth/FacebookAuthProvider.swift:2]: Can you please change these names as well after renaming the files? You might need to make this change in all the relevant files like  `SocialAuthView`, `SocialAuthViewModel` etc.

File: Core/Core/View/Base/SocialAuthButton.swift [hunk 42]
```diff
@@ -0,0 +1,74 @@
+//
+//  LabelButton.swift
```
Comment (saeedbashir) [inline Core/Core/View/Base/SocialAuthButton.swift:2]: Please change this as well

**Review by saeedbashir — CHANGES_REQUESTED** [review]

**Review by volodymyr-chekyrta — DISMISSED** [review]

**Review by saeedbashir — APPROVED** [review]
LGTM 👍 

Let the test cases pass.

**Review by volodymyr-chekyrta — APPROVED** [review]
